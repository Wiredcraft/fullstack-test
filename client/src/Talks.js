import React, { useState, useEffect, useContext } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Link } from '@reach/router';

import Talk from './Talk';
import Title from './Title';
import Spinner from './Spinner';
// import Pagination from './Pagination';

import config from './config';

const Talks = (props) => {
  const [ talks, setTalks ] = useState([]);
  const [ loaded, setLoaded ] = useState(false);
  const [ active, setActive ] = useState(null);

  // const [ length, setLength ] = useState(0);

  useEffect(() => {
    setLoaded(false);

    fetch(`${config.apiHost}/talks`)
    .then(async res => {
      const json = await res.json();

      const { talks, length } = json;

      const newTalks = Object.keys(talks)
        .map(k => {
          const t = talks[k];
          t.key = k;
          t.votes = t.votes || [];
          return t;
        });

      setTalks(newTalks);
      // setLength(length);

      setLoaded(true);
    });

  }, []);

  return (
    <React.Fragment>
      <Title text='Home' />
      {
        loaded
        ? (<div className='talks' onClickCapture={() => setActive(null)}>
          <div className='talks-inner body-container'>
            <div className='talks-header'>
              <h2>All Talks</h2>
              <Link to={`${process.env.PUBLIC_URL}/talks/new`} className='new-talk-btn'>
                <FontAwesomeIcon
                  icon='plus'
                  role='button'
                  aria-label='new talk'
                  title='new talk'
                />
              </Link>
            </div>
            <div>
              {
                talks.length
                  ? talks.map(({ key, ...talk}) => {
                    return (
                      <Talk
                        active={active === key}
                        key={key}
                        id={key}
                        clickHandler={() => setActive(key)}
                        {...talk}
                      />
                    );
                  })
                  : <p>
                    No talks yet.
                    <Link to={`${process.env.PUBLIC_URL}/talks/new`}>
                      Add one!
                    </Link>
                  </p>
              }
            </div>
          </div>
        </div>)
        : <Spinner />
      }
      {/*<Pagination
        prev={length}
        next={length}
      />*/}
    </React.Fragment>
  );
}

export default Talks;
