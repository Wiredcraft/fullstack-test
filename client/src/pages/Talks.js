import React, { useState, useEffect, useContext } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Link } from '@reach/router';

import TalkCard from '../components/TalkCard';
import Title from '../components/Title';
import Spinner from '../components/Spinner';
import UserContext from '../components/UserContext';

import { fireModal, NOT_LOGGED_IN, NETWORK_ERROR } from '../utils/fireModal';
import config from '../utils/config';

import  { LOADING, LOADED, ERROR } from '../utils/loadingStatuses';

const Talks = () => {
  const [ talks, setTalks ] = useState([]);
  const [ status, setStatus ] = useState(LOADING);
  const [ active, setActive ] = useState(null);

  const talkCardRef = React.createRef();

  const setVotes = (talkId, newVotes) => {
    const talk = talks.find(talk => talk.id === talkId);
    talk.votes = newVotes;
    setTalks(talks);
    const _current = talkCardRef.current;
    setActive(null);
    _current.focus();
    setActive(talkId);
  };

  const { user } = useContext(UserContext);

  useEffect(() => {
    setStatus(LOADING);

    (async () => {

      try {
        const res = await fetch(`${config.apiHost}/talks`);
        const json = await res.json();

        const { talks } = json;

        const newTalks = Object.keys(talks)
        .map(k => {
          const t = talks[k];
          t.id = k;
          return t;
        });

        setTalks(newTalks);

        setStatus(LOADED);
      } catch(e) {
        fireModal(NETWORK_ERROR);
        setStatus(ERROR);
      }

    })();

  }, []);

  const checkUser = e => {
    if (!user) {
      e.preventDefault();
      fireModal(NOT_LOGGED_IN, {actionName: 'create a new talk'});
    }
  };

  const orderByVotes = talks => talks.sort((a, b) => b.votes - a.votes);

  return (
    <React.Fragment>
      <Title text='Home' />
      {status === LOADING
        ? <Spinner />
        : (<div className='talks' onClickCapture={() => setActive(null)}>
            <div className='talks-inner body-container'>
            {status === ERROR
              ? (<React.Fragment>
                  <h2>Network Error</h2>
                  <p>Please try again later.</p>
                </React.Fragment>
              )
              : (<React.Fragment>
                <div className='talks-header'>
                  <h2>Upcoming Talk Proposals</h2>
                  <Link
                    to={`${process.env.PUBLIC_URL}/talks/new`}
                    className='new-talk-btn'
                    onClick={checkUser}
                  >
                    <FontAwesomeIcon
                      icon='plus'
                      role='button'
                      aria-label='new talk'
                      title='new talk'
                    />
                  </Link>
                </div>
                <div>
                {
                  talks.length
                  ? orderByVotes(talks)
                  .map(({ id, ...talk }, idx) => {
                    return (
                      <TalkCard
                        ref={talkCardRef}
                        idx={idx}
                        updateVotes={newVotes => {
                          setVotes(id, newVotes);
                        }}
                        active={active === id}
                        key={id}
                        id={id}
                        setActive={setActive}
                        {...talk}
                      />
                    );
                  })
                  : <p>
                    {'No talks yet. '}
                    <Link className='styled-link' to={`${process.env.PUBLIC_URL}/talks/new`}>
                     Add one!
                    </Link>
                  </p>
                }
              </div>
            </React.Fragment>)
            }

          </div>
        </div>)
      }
    </React.Fragment>
  );
}

export default Talks;
