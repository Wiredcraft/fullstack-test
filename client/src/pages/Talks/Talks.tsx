import { Route, Routes } from 'react-router-dom';

import { TalksList } from './TalksList';

export function Talks() {
  return (
    <Routes>
      <Route index element={<TalksList />} />
    </Routes>
  );
}
