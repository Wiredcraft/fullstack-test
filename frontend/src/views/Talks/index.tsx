import PropTypes from 'prop-types';
import React, {ReactElement} from 'react';
import {IAppState} from 'src/interfaces/IRootState';
import {api} from '../../actions';
import {connect} from 'react-redux';
import {getCookie} from '../../utils/cookies';

/**
  * Talks component.
  * @param {object} props - Component props
  * @return {ReactElement}
 */
function Talks(props: any): ReactElement {
  const abcd = (event: any): void => {
    event.preventDefault();

    props.dispatch(api.talks.apiTalksListAction(getCookie('token')));
  };


  return (
    <div>
        Welcome, {getCookie('username')} !<p/>
      <button onClick={abcd}>List talks</button>
    </div>
  );
}

Talks.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const mapStateToProps = (_state: IAppState) => {
  return {};
};

export default connect(mapStateToProps)(Talks);
