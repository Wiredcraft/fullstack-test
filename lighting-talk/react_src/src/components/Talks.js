import React, {Component} from 'react';
import Talk from './Talk';
import PrimaryMenu from "./PrimaryMenu";
import axios from 'axios';
class Talks extends Component {
    constructor() {
        super();
        this.state = {
            talks: [],
            msgType: '',
            msgHeader: '',
            msgContent: '',
            msgShow: false
        }
    }

    getTalks() {
        this.setState({talks: []});
        axios.get('/api/talks?filter[order]=voteAverage%20DESC')
        .then(response => {
            this.setState({talks: response.data})
        })
        .catch(err => console.log(err));
    }

    componentWillMount() {
        this.getTalks();
    }

    showMessage(messageItem){
        this.setState({msgType: messageItem.type, msgHeader: messageItem.header, msgContent: messageItem.content, msgShow: true})
        this.getTalks();
    }

    render() {
        const talkItems = this.state.talks.map((talk,i) => {
            return(
                <div key={talk.id}><Talk key={talk.id} item={talk} callback={(messageItem) => this.showMessage(messageItem)}/><br/></div>
            )
        })
        return (
            <div className="ui raised container segment">
                <h1 className="ui header">Lighting Talks</h1>
                <div className={"ui " + (this.state.msgShow ? 'visible ' : 'hidden ') + this.state.msgType + " message"}>
                    <div className="header">
                        {this.state.msgHeader}
                    </div>
                    <p>{this.state.msgContent}</p>
                </div>
                <PrimaryMenu callback={(messageItem) => this.showMessage(messageItem)} />
                <br />
                {talkItems}
            </div>
        )
    }
}

export default Talks;