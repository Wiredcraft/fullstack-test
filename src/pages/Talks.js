import React, {Component} from 'react'
import {connect} from 'react-redux'
import TalkBox from '../components/TalkBox'
import TalkList from '../components/TalkList'
import {fetchTalks, voteForTalk} from '../redux/talks'
import {changeField, submitBox} from '../redux/talkBox'

const mapStateToProps = state => {
  return {
    talks: state.talks.sort((talkA, talkB) => talkB.votes - talkA.votes),
    talkBox: state.talkBox,
  }
}

class Talks extends Component {
  componentDidMount() {
    this.props.fetchTalks()
  }

  render() {
    const {talks, talkBox, voteForTalk, submitBox, changeField} = this.props

    return (
      <div className="Talks">
        <TalkList
          talks={talks}
          onVoteForTalk={voteForTalk}
        />
        <TalkBox
          talkBox={talkBox}
          onChangeField={changeField}
          onSubmitBox={submitBox}
        />
      </div>
    )
  }
}

export default connect(mapStateToProps, {
  fetchTalks,
  voteForTalk,
  submitBox,
  changeField,
})(Talks)
