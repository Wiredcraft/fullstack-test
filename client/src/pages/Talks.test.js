import React from 'react';
import { render } from 'react-testing-library';
import Talks from './Talks';

it('renders Talks correctly', () => {
  const { asFragment } = render(<Talks />);

  expect(asFragment()).toMatchSnapshot();
});
