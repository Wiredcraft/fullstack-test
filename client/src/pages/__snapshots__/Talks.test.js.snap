// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Talks correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="spinner-container"
    >
      <div
        aria-label="Loading..."
        class="lds-ring"
        title="Loading..."
      >
        <div
          class="lds-ring-inner first"
        />
        <div
          class="lds-ring-inner second"
        />
        <div
          class="lds-ring-inner third"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
